llvm/lib/Target/X86/X86ISelLowering.cpp

```diff
@@ -59668,3 +59668,19 @@ Align X86TargetLowering::getPrefLoopAlignment(MachineLoop *ML) const {
     return Align(1ULL << ExperimentalPrefInnermostLoopAlignment);
   return TargetLowering::getPrefLoopAlignment();
 }
+
+bool X86TargetLowering::isDesirableToCommuteWithShift(
+    const SDNode *N, CombineLevel Level) const {
+  using namespace llvm::SDPatternMatch;
+  assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
+          N->getOpcode() == ISD::SRL) &&
+         "Expected shift op");
+
+  SDValue ShiftLHS = N->getOperand(0);
+  SDValue Add;
+  if (sd_match(ShiftLHS, m_OneUse(m_SExt(m_OneUse(m_Value(Add))))) ||
```

Comment:
The rest of the helpers should use the sd variant.

Reply:
fixed

Reply:
This does not seem done...

Reply:
Some architectures, such as ARM, use this logic（sd_match(ShiftLHS, m_OneUse(m_SExt(m_OneUse(m_Value(Add)） to directly return true, which doesn't seem appropriate, so I didn't change it.